components/feature/downloads/src/main/java/mozilla/components/feature/downloads/share/ShareDownloadMiddleware.kt

```diff
@@ -0,0 +1,180 @@
+import mozilla.components.support.ktx.android.content.shareMedia
+import mozilla.components.support.ktx.kotlin.sanitizeURL
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import kotlin.math.absoluteValue
+import kotlin.random.Random
+
+/**
+ * 200 response code for a successful network request.
+ */
+@VisibleForTesting
+internal const val OK_STATUS = 200
+
+/**
+ * At most time to allow fo the file to be downloaded and the share intent chooser to appear.
+ */
+private const val OPERATION_TIMEOUT_MILLIS = 1000L
+
+/**
+ * Subdirectory of Context.getCacheDir() where the resources to be shared are stored.
+ *
+ * Location must be kept in sync with the paths our FileProvider can share from.
+ */
+@VisibleForTesting
+internal var cacherDirName = "mozac_share_cache"
+
+/**
+ * [Middleware] implementation for sharing online resources.
+ *
+ * This will intercept only [DownloadAction.AddDownloadAction] [BrowserAction]s
+ * with the source set to be [DownloadState.Source.SHARE].
+ *
+ * Following which it will transparently
+ *  - download internet resources while respecting the private mode related to cookies handling
+ *  - temporarily cache the downloaded resources
+ *  - automatically open the platform app chooser to share the cached files with other installed Android apps.
+ *
+ * with a 1 second timeout to ensure a smooth UX.
+ *
+ *  @param context Android context used for vaious platform interactions
+ *  @param httpClient Client used for downloading internet resources
+ *  @param scope CoroutineScope used for background work execution
+ */
+class ShareDownloadMiddleware(
+    private val context: Context,
+    private val httpClient: Client,
+    private val scope: CoroutineScope = CoroutineScope(IO)
+) : Middleware<BrowserState, BrowserAction> {
+    @VisibleForTesting
+    internal val logger = Logger("ShareDownloadMiddleware")
+
+    init {
+        scope.launch {
+            cleanupCache()
+        }
+    }
+
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        if (action is DownloadAction.AddDownloadAction &&
+            action.download.source == DownloadState.Source.SHARE
+        ) {
+            logger.debug("Starting the sharing process")
+            startSharing(action.download)
+        } else {
+            next(action)
+        }
+    }
+
+    @VisibleForTesting
+    internal fun startSharing(fileToShare: DownloadState) {
+        val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+            when (throwable) {
+                is InterruptedException -> {
+                    logger.warn("Share failed: operation timeout reached")
+                }
+                is IOException,
+                is RuntimeException,
+                is NullPointerException -> {
+                    logger.warn("Share failed: $throwable")
+                }
```

Comment:
Should we should some UI to inform the user that something unexpected happened while sharing?

Reply:
When downloading we use a toast to make the user aware of some errors https://github.com/mozilla-mobile/android-components/blob/c1a85a4254a7a79eab60fe3c71a67e15f7db101f/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L232

Reply:
Saw Chrome failing silently but maybe we could do better.
Asked UX in the ticket.

Reply:
👍🏽